L46-core-tls-credential-API.md

```diff
@@ -0,0 +1,425 @@
+// certain security features supported in TLS.
+class TlsCredentialsOptions {
+ public:
+  ~TlsCredentialsOptions();
+
+  // Copy constructor does a deep copy. No assignment permitted.
+  TlsCredentialsOptions(const TlsCredentialsOptions& other);
+
+  TlsCredentialsOptions& operator=(const TlsCredentialsOptions& other) = delete;
+
+  // ---- Setters for member fields ----
+  // Sets the certificate provider used to store root certificates and the
+  // certificate chain conveying the application's identity and corresponding
+  // private key.
+  void set_certificate_provider(
+      std::shared_ptr<CertificateProviderInterface> certificate_provider);
+
+  // Watches the updates of root certificates with name |root_cert_name| (set by
+  // the user via set_root_cert_name).  If used in TLS credentials, setting this
+  // field is optional for both the client side and the server side.
+  // If this is not set on the client side, we will use the root certificates
+  // stored in the default system location, since client side must provide root
+  // certificates in TLS (no matter single-side TLS or mutual TLS).
+  // If this is not set on the server side, we will not watch any root
+  // certificate updates, and assume no root certificates needed for the server
+  // (in the one-side TLS scenario, the server is not required to provide root
+  // certificates). We don't support default root certs on server side.
+  void watch_root_certificates();
+
+  // Sets the name of root certificates being watched, if |watch_root_certificates| is
+  // called. If not set, an empty string will be used as the name. For simple
+  // use cases this is not needed, but a certificate provider could be
+  // implemented that watches many sets of certificates. This allows the user to
+  // differentiate between different root certificates.
+  // 
+  // @param root_cert_name the name of root certificates being set.
+  void set_root_cert_name(const std::string& root_cert_name);
+
+  // Watches the updates of identity key-certificate pairs with name
+  // |identity_cert_name|. If used in TLS credentials, it is required to be set
+  // on the server side, and optional for the client side(in the one-side
+  // TLS scenario, the client is not required to provide identity certificates).
+  void watch_identity_key_cert_pairs();
+
+  // Sets the name of identity key-certificate pairs being watched, if
+  // |watch_identity_key_cert_pairs| is called. If not set, an empty string will
+  // be used as the name.
+  //
+  // @param identity_cert_name the name of identity key-certificate pairs being set.
+  void set_identity_cert_name(const std::string& identity_cert_name);
+
+  // Sets the TLS session key logging file path. If not set, TLS
+  // session key logging is disabled. Note that this should be used only for
+  // debugging purposes. It should never be used in a production environment
+  // due to security concerns.
```

Comment:
nit: Can we be more specific than "due to security concerns"? For example, we can say "because anyone who can obtain the (logged) session key can decrypt all traffic on a connection"?

Reply:
Done